modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotPartitionsVerifyTask.java

```diff
@@ -0,0 +1,254 @@
+        @IgniteInstanceResource
+        private IgniteEx ignite;
+
+        /** Injected logger. */
+        @LoggerResource
+        private IgniteLogger log;
+
+        /** Snapshot name to validate. */
+        private String snpName;
+
+        /** Consistent snapshot metadata file name. */
+        private String consId;
+
+        /**
+         * @param snpName Snapshot name to validate.
+         * @param consId Consistent snapshot metadata file name.
+         */
+        public VisorVerifySnapshotPartitionsJob(String snpName, String consId) {
+            this.snpName = snpName;
+            this.consId = consId;
+        }
+
+        @Override public Map<PartitionKeyV2, PartitionHashRecordV2> execute() throws IgniteException {
+            IgniteSnapshotManager snpMgr = ignite.context().cache().context().snapshotMgr();
+
+            if (log.isInfoEnabled()) {
+                log.info("Verify snapshot partitions procedure has been initiated " +
+                    "[snpName=" + snpName + ", consId=" + consId + ']');
+            }
+
+            SnapshotMetadata meta = snpMgr.readSnapshotMetadata(snpName, consId);
+            Set<Integer> grps = new HashSet<>(meta.partitions().keySet());
+            Set<T2<File, File>> pairs = new HashSet<>();
+
+            for (File dir : snpMgr.snapshotCacheDirectories(snpName, consId)) {
+                int grpId = CU.cacheId(cacheGroupName(dir));
+
+                if (!grps.remove(grpId))
+                    continue;
+
+                Set<Integer> parts = new HashSet<>(meta.partitions().get(grpId));
+
+                for (File part : cachePartitions(dir)) {
+                    int partId = partId(part.getName());
+
+                    if (!parts.remove(partId))
+                        continue;
+
+                    pairs.add(new T2<>(dir, part));
+                }
+
+                if (!parts.isEmpty()) {
+                    throw new IgniteException("Snapshot data doesn't contain required cache group partition " +
+                        "[grpId=" + grpId + ", snpName=" + snpName + ", consId=" + consId +
+                        ", missed=" + parts + ", meta=" + meta + ']');
+                }
+            }
+
+            if (!grps.isEmpty()) {
+                throw new IgniteException("Snapshot data doesn't contain required cache groups " +
+                    "[grps=" + grps + ", snpName=" + snpName + ", consId=" + consId +
+                    ", meta=" + meta + ']');
+            }
+
+            Map<PartitionKeyV2, PartitionHashRecordV2> res = new HashMap<>();
+            ThreadLocal<ByteBuffer> buff = ThreadLocal.withInitial(() -> ByteBuffer.allocateDirect(meta.pageSize())
+                .order(ByteOrder.nativeOrder()));
+
+            try {
+                U.doInParallel(
+                    ignite.context().getSystemExecutorService(),
+                    pairs,
+                    pair -> {
+                        String grpName = pair.get1().getName();
```

Comment:
`get1()` is group dir, not group name. Looks like you don't need group dir here at all, you should store group name. Also, group dir can be obtained from partition file using `getParent()` method, it will be easier to read the code if there will be collection of part files, but not collection of some pairs.

Reply:
Fixed.